### PR TITLE
Derivative operations in Builder

### PR DESCRIPTION
**Pull request by trenouf**
This commit adds the derivative operations to Builder, and switches the
SPIR-V reader to use Builder for them.

I also removed an unnecessary wqm intrinsic call in the middle of the
calculation.

Change-Id: Ie71ba1d9025ffc9e0c55509c80c47b4bcbcdb401

### CODE REVIEW

**Comment by trenouf** [comment]
This PR contains two commits, because it depends on an earlier still-open PR. Please review only the topmost commit.

**Comment by linqun** [comment]
I will review it once the test is passed. please let me know if this failure is due to Jenkins agent.

**Comment by trenouf** [comment]
This looks like a spurious failure. But I will rebase it before retriggering testing.

**Comment by trenouf** [comment]
V2: Rebased.

File: builder/llpcBuilderImplArith.cpp [hunk 1]
```diff
@@ -526,6 +526,117 @@ Value* BuilderImplArith::CreateFSign(
...
+    bool          isY,        // False for derivative in X direction, true for Y direction
+    bool          isFine,     // True for "fine" calculation, where the value in the current fragment is used.
+                              // False for "coarse" calculation, where it might use fewer locations to calculate.
+    const Twine&  instName)   // [in] Name to give instruction(s)
+{
+    uint32_t tableIdx = isY * 2 + isFine;
+    Value* pResult = nullptr;
+    if (SupportDpp())
+    {
+        // DPP (gfx8+) version.
```
Comment (amdrexu) [inline builder/llpcBuilderImplArith.cpp]: gfx8+ -> GFX8+

File: builder/llpcBuilderImplArith.cpp [hunk 2]
```diff
@@ -526,6 +526,117 @@ Value* BuilderImplArith::CreateFSign(
...
+                                                                   });
+                               pSecondVal = CreateZExtOrTrunc(pSecondVal, getIntNTy(pValTy->getPrimitiveSizeInBits()));
+                               pSecondVal = CreateBitCast(pSecondVal, pValTy);
+                               Value* pResult = CreateFSub(pFirstVal, pSecondVal);
+                               return CreateUnaryIntrinsic(Intrinsic::amdgcn_wqm, pResult);
+                            });
+    }
+    else
+    {
+        // ds_swizzle (pre-gfx8) version
```
Comment (amdrexu) [inline builder/llpcBuilderImplArith.cpp]: gfx8 -> GFX8

File: builder/llpcBuilderImplArith.cpp [hunk 3]
```diff
@@ -526,6 +526,117 @@ Value* BuilderImplArith::CreateFSign(
     return CreateSelect(pIsNonNegative, pPartialResult, ConstantFP::get(pInValue->getType(), -1.0), instName);
 }
 
+// =====================================================================================================================
+// Create derivative calculation on float or vector of float or half
+Value* BuilderImplArith::CreateDerivative(
+    Value*        pInValue,   // [in] Input value
```
Comment (amdrexu) [inline builder/llpcBuilderImplArith.cpp]: pInValue -> pValue

File: builder/llpcBuilderImplArith.cpp [hunk 4]
```diff
@@ -526,6 +526,117 @@ Value* BuilderImplArith::CreateFSign(
...
+    bool          isY,        // False for derivative in X direction, true for Y direction
```
Comment (amdrexu) [inline builder/llpcBuilderImplArith.cpp]: isY -> isDirectionY
I think it is necessary to make it more accurate

File: builder/llpcBuilderRecorder.cpp [hunk 5]
```diff
@@ -648,6 +650,17 @@ Value* BuilderRecorder::CreateSMod(
     return Record(Opcode::SMod, pDividend->getType(), { pDividend, pDivisor }, instName);
 }
 
+// =====================================================================================================================
+// Create derivative calculation on float or vector of float or half
+Value* BuilderRecorder::CreateDerivative(
+    Value*        pInValue,   // [in] Input value
+    bool          isY,        // False for derivative in X direction, true for Y direction
```
Comment (amdrexu) [inline builder/llpcBuilderRecorder.cpp]: isY -> isDirectionY

File: builder/llpcBuilderRecorder.cpp [hunk 6]
```diff
@@ -648,6 +650,17 @@ Value* BuilderRecorder::CreateSMod(
     return Record(Opcode::SMod, pDividend->getType(), { pDividend, pDivisor }, instName);
 }
 
+// =====================================================================================================================
+// Create derivative calculation on float or vector of float or half
+Value* BuilderRecorder::CreateDerivative(
+    Value*        pInValue,   // [in] Input value
```
Comment (amdrexu) [inline builder/llpcBuilderRecorder.cpp]: pInValue -> pValue

File: builder/llpcBuilderRecorder.h [hunk 7]
```diff
@@ -237,6 +238,9 @@ class BuilderRecorder final : public Builder, BuilderRecorderMetadataKinds
     Value* CreateLog(Value* pInValue, const Twine& instName = "") override final;
     Value* CreateInverseSqrt(Value* pInValue, const Twine& instName = "") override final;
 
+    // Create derivative calculation on float or vector of float or half
+    Value* CreateDerivative(Value* pInValue, bool isY, bool isFine, const Twine& instName = "") override final;
```
Comment (amdrexu) [inline builder/llpcBuilderRecorder.h]: pInValue -> pValue

**Comment by trenouf** [comment]
Qun also said in #166 that he wants CreateDerivative in the misc group.

**Comment by amdrexu** [comment]
I am OK with moving it to misc group. I think Qun said this becasuse cross-lane operations are involved and it is not simple arithmetic operations any more.

**Comment by trenouf** [comment]
V3: Rebased and addressed review comments, including moving
        CreateDerivative into misc group, and thus moving Scalarize() helper
        functions into BuilderImplBase.

This is now a single commit.

**Comment by trenouf** [comment]
I will land this.

**Comment by amdvlk-admin** [comment]
### Test summary for commit cf38d02661778796933439b522eb8d3d8b377409
<details open><summary>Driver commits used in build</summary><ul>

<li>METROHASH: amd-mainline 2b6fee002db6cc92345b02aeee963ebaaf4c0e2f</li>
<li>PAL: dev 68b57dba33a4d922e8f1ef1b3781c2f659ffbd1c</li>
<li>SPVGEN: dev 2f31d1170e8a12a66168b23235638c4bbc43ecdc</li>
<li>XGL: dev 331558e93794068a786bf699d3fe23bb11bac021</li>
<li>LLVM: amd-vulkan-dev f0a8af5e93d57b890687c7ced43943a0c3a167ca</li>
</ul></details>

<details open><summary>ShaderDB test</summary><ul>

<li>Passed: 925/925 (100.0%)</li>
<li>Failed: 0/925 (0.0%)</li>
</ul></details>

<details open><summary>CTS tests (built with branch Vulkan-CTS-1.1.2.3-cherry-picks)</summary><ul>


<details open><summary>Ubuntu 18.04, Ellesmere</summary><ul>

<li>Passed: 270023/270024 (100.0%)</li>
<li><details><summary>Failed: 1/270024 (0.0%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.renderpass2.dedicated_allocation.attachment_allocation.input_output.91
Stack trace: Timeout or crahsed
```
</p></details></li>


</ul></details>

<details open><summary>Rhel 7.6, Vega10</summary><ul>

<li>Passed: 366376/366376 (100.0%)</li>
<li>Failed: 0/366376 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by trenouf** [comment]
retest this please

**Comment by amdvlk-admin** [comment]
### Test summary for commit cf38d02661778796933439b522eb8d3d8b377409
<details open><summary>Driver commits used in build</summary><ul>

<li>METROHASH: amd-mainline 2b6fee002db6cc92345b02aeee963ebaaf4c0e2f</li>
<li>PAL: dev 68b57dba33a4d922e8f1ef1b3781c2f659ffbd1c</li>
<li>SPVGEN: dev 2f31d1170e8a12a66168b23235638c4bbc43ecdc</li>
<li>XGL: dev 331558e93794068a786bf699d3fe23bb11bac021</li>
<li>LLVM: amd-vulkan-dev f0a8af5e93d57b890687c7ced43943a0c3a167ca</li>
</ul></details>

<details open><summary>ShaderDB test</summary><ul>

<li>Passed: 925/925 (100.0%)</li>
<li>Failed: 0/925 (0.0%)</li>
</ul></details>

<details open><summary>CTS tests (built with branch Vulkan-CTS-1.1.2.3-cherry-picks)</summary><ul>


<details open><summary>Ubuntu 18.04, Ellesmere</summary><ul>

<li>Passed: 366250/366250 (100.0%)</li>
<li>Failed: 0/366250 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 7.6, Vega10</summary><ul>

<li>Passed: 282963/282963 (100.0%)</li>
<li>Failed: 0/282963 (0.0%)</li>
</ul></details>


</ul></details>

**Pull request merged**